### PR TITLE
Yosys Synthesis Script Cleanup

### PR DESCRIPTION
**Pull request by donn**
* `Yosys.*`
  * Fixes a bug where synthesis checks were not passed properly when `SYNTH_ELABORATE_ONLY` is true. 
  * **Internal**:
    * Created new namespace, `yosys_ol`, to encapsulate a number of reusable functions for modularity and readability
    * Created two new functions `ol_proc` and `ol_synth`; to encapsulate our modified `synth` and `proc` instead of them being strewn across `synthesize.tcl`
    * ABC script construction moved to standalone file

### CODE REVIEW

File: openlane/scripts/yosys/json_header.tcl [hunk 1]
```diff
@@ -13,12 +13,11 @@
...
 source $::env(SCRIPTS_DIR)/yosys/common.tcl
-set vtop $::env(DESIGN_NAME)
 
 source $::env(_DEPS_SCRIPT)
 
-read_verilog_files $vtop
-hierarchy -check -top $vtop
-yosys rename -top $vtop
+yosys_ol::read_verilog_files $::env(DESIGN_NAME)
+hierarchy -check -top $::env(DESIGN_NAME) -nokeep_prints -nokeep_asserts
```
Comment (kareefardi) [inline openlane/scripts/yosys/json_header.tcl:20]: what are `-nokeep_prints` and `-nokeep_asserts` for?

File: openlane/scripts/yosys/synthesize.tcl [hunk 2]
```diff
@@ -11,225 +11,75 @@
...
-    set dfflib $::env(SYNTH_LIBS)
 }
 
 set dfflib_args [list]
 foreach lib $dfflib {
     lappend dfflib_args -liberty $lib
 }
 
+set dfflib $::env(SYNTH_LIBS)
+if {[info exists ::env(DFF_LIB_SYNTH)]} {
```
Comment (kareefardi) [inline openlane/scripts/yosys/synthesize.tcl:41]: this is repeated in line 31

File: openlane/scripts/yosys/synthesize.tcl [hunk 3]
```diff
@@ -248,120 +98,73 @@ if { $::env(SYNTH_ELABORATE_ONLY) } {
...
+set adder_type $::env(SYNTH_ADDER_TYPE)
+if { !($adder_type in [list "YOSYS" "FA" "RCA" "CSA"]) } {
+    log -stderr "\[ERROR] Misformatted SYNTH_ADDER_TYPE (\"$::env(SYNTH_ADDER_TYPE)\")."
+    log -stderr "\[ERROR] Correct format is \"YOSYS|FA|RCA|CSA\"."
+    exit 1
+}
 if { $adder_type == "RCA"} {
     if { [info exists ::env(SYNTH_RCA_MAP)] } {
-        log "\[INFO] Applying ripple carry adder mapping from '$::env(RIPPLE_CARRY_ADDER_MAP)'…"
+        log "\[INFO] Applying ripple carry adder mapping from '$::env(RIPPLE_CARRY_ADDER_MAP)'..."
```
Comment (kareefardi) [inline openlane/scripts/yosys/synthesize.tcl:116]: why?

File: openlane/scripts/yosys/common.tcl [hunk 4]
```diff
@@ -11,44 +11,111 @@
...
+        proc_prune
+        proc_init
+        proc_arst
+        proc_rom
+        proc_mux
+        tee -o "$report_dir/latch.rpt" proc_dlatch
+        proc_dff
+        proc_memwr
+        proc_clean
+        opt_expr -keepdc
```
Comment (kareefardi) [inline openlane/scripts/yosys/common.tcl:45]: check was previously just after `proc_clean`
also we didn't have `-keepdc`

File: openlane/scripts/yosys/json_header.tcl (hunk 1 above)
Comment (donn) [reply openlane/scripts/yosys/json_header.tcl:20]: Allegedly, they remove print and assert cells that cause issues in OpenROAD. Practically, nothing, but they don't hurt either.

File: openlane/scripts/yosys/synthesize.tcl (hunk 3 above)
Comment (donn) [reply openlane/scripts/yosys/synthesize.tcl:116]: Yosys appears to have a problem with Unicode

**Review by kareefardi — CHANGES_REQUESTED** [review]
Should fix at least the following changed design (since it now has hold violations):
- `MS_SPI_XIP_CACHE`

Wondering about increased `synthesis__check_error__count` in `aes_user_project_wrapper`, `caravel_upw` and `io_placer`

**Review by kareefardi — APPROVED** [review]

**Pull request merged**